sympy/functions/special/bessel.py

```diff
@@ -232,6 +232,8 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         if e.is_positive:
             return arg**nu/(2**nu*gamma(nu + 1))
         elif e.is_negative:
+            # Refer https://functions.wolfram.com/Bessel-TypeFunctions/BesselJ/06/02/02/02/0005/
+            # for more information on asymptotic approximation of besselj function.
```

Comment:
This does not seem agree with Wikipedia. It looks like wolfram has a piecewise expression and this is only one of its parts.

Reply:
Yeah correct but the code ensures the we would only need one branch for all bessel functions and that is the branch which includes `arg(z) == 0` . I had explained this in a comment of mine (https://github.com/sympy/sympy/pull/23844#discussion_r945219091) . As we now use 
```
        if z0.is_infinite:
            cdir = sign(z0)
            cdir = cdir/abs(cdir)
            e = e.subs(z, cdir*z)
            dir = "-"
            z0 = S.Infinity
```
We don't need to care about different branches where `abs(z) -> oo` , we just need to care about `z -> oo` which is given by `arg(z) == 0` (as `arg(oo) == 0`) . Hence expression for any `abs(z) -> oo`  will be transformed to some expression such that `z -> oo` and then enter the `_eval_as_leading_term` block with a `cdir = 1` . Hence we just need the branch handling `oo`. I've added tests for all kind of infinities and everything works perfectly fine .
```
    assert limit(besselj(1, x), x, oo) == 0
    assert limit(besselj(1, x), x, -oo) == 0
    assert limit(besselj(1, x), x, I*oo) == oo*I
    assert limit(besselj(1, x), x, -I*oo) == -oo*I
```

Reply:
Therefore similarly for all `bessel` functions , we just need to make sure we correctly add the branch including `z->oo` or `arg(z) == 0`

Reply:
Are you sure that this expression agrees with that of [Wikipedia](https://en.wikipedia.org/wiki/Bessel_function#Asymptotic_forms) as `z -> oo`?

Reply:
Yeah .... though we can go for something simpler or rather similar to what wikipedia says in terms of `cos`
<img width="444" alt="image" src="https://user-images.githubusercontent.com/87052487/185747590-bf8f553c-aa59-4038-9493-25d22916c963.png">

Wolfram presents the asymptotic approximation in two forms
<img width="297" alt="image" src="https://user-images.githubusercontent.com/87052487/185747626-3c24766a-7f1c-4450-b3f8-0d00b6e31db5.png">

And like we can find the exact corresponding approximation based on trigonometric functions . 

<img width="234" alt="image" src="https://user-images.githubusercontent.com/87052487/185747666-c8e7a0ca-8ec6-4892-ab92-9c78392f48b0.png">

I see that the `cos` form is more established than the other one in form of exponentials. So yeah for the one going to `oo`, we'll be using the first one of this . I said well established , because I could find a couple references for proofs of this form in an e-book and also on maths stackoverflow . Here's the [proof](https://math.stackexchange.com/questions/177118/about-the-asymptotic-formula-of-bessel-function) for the approximation. Though as I said at the end both are corresponding forms .I tried limits from both and I get the same results 
```
>>> limit(besselj(1, x), x, oo)
0
>>> limit(besselj(1, x), x, I*oo)
oo*I
>>> limit(besselj(2, x), x, I*oo)
-oo
>>> limit(besselj(3, x), x, I*oo)
-oo*I
>>> limit(besselj(4, x), x, I*oo)
oo
>>> limit(besselj(5, x), x, I*oo)
oo*I
```

Reply:
I am fully sure about the modified ones ( because these don't really have trigo forms for their approximations) but if we want to make it simpler or according to what wolfram says (using `sin` and `cos`) we can use this diff
```
--- a/sympy/functions/special/bessel.py
+++ b/sympy/functions/special/bessel.py
@@ -234,8 +234,8 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         elif e.is_negative:
             # Refer https://functions.wolfram.com/Bessel-TypeFunctions/BesselJ/06/02/02/02/0005/
             # for more information on asymptotic approximation of besselj function.
-            return sqrt(2)*(-(-1)**(S(3)/4)*I**nu*exp(-I*pi*nu + I*z) + \
-                    (-1)**(S(1)/4)*I**nu*exp(-I*z))/(2*sqrt(pi)*sqrt(z))
+            return sqrt(2)*cos(z - pi*(2*nu + 1)/4)/sqrt(pi*z)

         return super(besselj, self)._eval_as_leading_term(x, logx, cdir)

@@ -365,7 +365,7 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         elif e.is_negative:
             # Refer https://functions.wolfram.com/Bessel-TypeFunctions/BesselY/06/02/02/02/0006/
             # for more information on asymptotic approximation of bessely function
-            return (-I*exp(-I*pi*nu/2 + I*z - I*pi/4) + I*exp(I*pi*nu/2 - I*z + I*pi/4))/sqrt(2*pi*z)
+            return sqrt(2)*sin(z - pi*nu/2 - pi/4)/sqrt(pi*z)

Reply:
Yupp , the expansions for `besselj` and `besseli` based on sin and cos are established as shown in Abramowitz and Stegun pg 364

<img width="333" alt="image" src="https://user-images.githubusercontent.com/87052487/185774528-e9526d55-e3dd-4cc7-a45a-dd8d1700fdd6.png">

So probably we can stick with these

Reply:
And for the modified bessel functions , these are also established(on pg 377 and 378) in forms of **exponentials** , so our current implementation is perfectly correct for these
<img width="381" alt="image" src="https://user-images.githubusercontent.com/87052487/185774617-9a1a07bf-a2c6-4b61-aa19-15dbe39a2cf2.png">

<img width="364" alt="image" src="https://user-images.githubusercontent.com/87052487/185774626-b35ef75f-2ad5-4c58-8686-44d891c18737.png">

Reply:
I like the expressions of Abramowitz and Stegun because they are obviously real for real z and nu. It is not clear that Wolframs expression have that property.

Reply:
Yeah and I realize that now we we're dealing with infinites as 
```
        if z0.is_infinite:
            cdir = sign(z0)
            cdir = cdir/abs(cdir)
            e = e.subs(z, cdir*z)
            dir = "-"
            z0 = S.Infinity
```
We only need to take care of `z -> oo` and not `abs(z) -> oo` . Hence one of your comments on the gitter channel can now we properly addressed. For asymptotic expansion of besseli you said
> I*exp(I*pi*nu - z) may be needed for some expansions if the dominating term is canceled but it does not seem to be a part of the leading term.

This was important for other infinites but not for `oo`, hence all limits shall now pass if we go with what `Abramowitz and Stegun ` say rather than what `Wolfram` says 
```
         elif e.is_negative:
             # Refer Abramowitz and Stegun 1965, p. 377 for more information on
             # asymptotic approximation of besseli function.
-            return (exp(z) + I*exp(I*pi*nu - z))/sqrt(2*pi*z)
+            return exp(z)/sqrt(2*pi*z)
```
This should be good enough

Reply:
I am quite satisfied with all 4 asymptotic expansions used (we have references for all) and particularly all expressions are built around `z->oo` . Let me know what you think !

I would also like to know if you can help me out with getting the values of something like `bessel(n, I*oo)`, where n is integer. 
For eg
```
>>> limit(bessely(1, x), x, I*oo)
bessely(1, oo*I)
>>> limit(bessely(2, x), x, I*oo)
bessely(2, oo*I)
```
These answers aren't wrong but if we could evaluate them then it would obviously be better .
For eg checkout `besselj`, it forms a repeated chain
```
>>> besselj(1, I*oo)
oo*I
>>> besselj(2, I*oo)
-oo
>>> besselj(3, I*oo)
-oo*I
>>> besselj(4, I*oo)
oo
>>> besselj(5, I*oo)
oo*I
>>> # repeats
```
This is made possible through 
```
            newz = z.extract_multiplicatively(I)
            if newz:  # NOTE we don't want to change the function if z==0
                return I**(-nu)*besselj(nu, -newz)
```
It seems `bessely` also forms such a cycle and repeats after where 4 integers. Although I have been trying to find some reference for this like the ones we have for `besselj`, I fail to find one for `bessely`.

Reply:
For `bessely`, it might be possible to use https://dlmf.nist.gov/10.17.E4

Reply:
We might not need that though, like the link points out a proper asymptotic series which can be implemented . In this case for leading term a good asymptotic approximation would suffice I would say . And if we analyze this a bit , `aseries` for all `bessel` functions , will be of the form `exp(z)*(Sum(f(k, nu), (k, 0, oo)))` . 

And like this was one of the main reasons we switched from 1st approach to 2nd approach ( approaches described below)

Approach 1) Using `aseries` and `gruntz` but however we try to mold it , this block in `mrv_leadterm` will raise a `PoleError` due to `exp(1/x)`
```
    w = Dummy("w", positive=True)
    f, logw = rewrite(exps, Omega, x, w)
    try:
        lt = f.leadterm(w, logx=logw)
    except (NotImplementedError, PoleError, ValueError):
        n0 = 1
        _series = Order(1)
        incr = S.One
        while _series.is_Order:
            _series = f._eval_nseries(w, n=n0+incr, logx=logw)
            incr *= 2
        series = _series.expand().removeO()
        try:
            lt = series.leadterm(w, logx=logw)
        except (NotImplementedError, PoleError, ValueError):
            lt = f.as_coeff_exponent(w)
            if lt[0].has(w):
                base = f.as_base_exp()[0].as_coeff_exponent(w)
                ex = f.as_base_exp()[1]
                lt = (base[0]**ex, base[1]*ex)
    return (lt[0].subs(log(w), logw), lt[1])
```
 
Approach 2) Using an asymptotic approximation in leading term and getting the result through this block in `limits.py` (last part)
```
        try:
            coeff, ex = newe.leadterm(z, cdir=cdir)
        except (ValueError, NotImplementedError, PoleError):
            # The NotImplementedError catching is for custom functions
            from sympy.simplify.powsimp import powsimp
            e = powsimp(e)
            if e.is_Pow:
                r = self.pow_heuristics(e)
                if r is not None:
                    return r
            try:
                coeff = newe.as_leading_term(z, cdir=cdir)
                if coeff != newe and coeff.has(exp):
                    return gruntz(coeff, z, 0, "-" if re(cdir).is_negative else "+")
            except (ValueError, NotImplementedError, PoleError):
                pass
```
So like `aseries` for let's take `besseli` looks something like this 
<img width="190" alt="image" src="https://user-images.githubusercontent.com/87052487/185842556-3997650f-e298-46a3-b275-b18d6b2e60e7.png">

Now implementing this isn't like to tedious or something and we can get it through 
```
    def _eval_aseries(self, n, args0, x, logx):
        from sympy.series.order import Order
        point = args0[1]
        nu, z = self.args

        def polynomial_coeff(n, v):
            if n == 0:
                return 1
            else:
                res = (-1)**n*cos(pi*v)*(gamma(n + S.Half + v)*
                    gamma(n + S.Half - v))/(pi* 2**n *gamma(n + 1))
                return res

        # Expansion at oo
        if point is S.Infinity:
            try:
                _, ex = z.subs(x, 1/x).leadterm(x)
            except (ValueError, NotImplementedError):
                return self

            if exp.is_positive and nu.is_integer:
                newn = ceiling(n/ex)                
                series = []

                for k in range(0, (newn + 1)//2):
                    term = polynomial_coeff(k, nu)/z**k
                    if k% 2 == 0:
                        series.append(term/sqrt(2*pi*z))
                    else:
                        series.append(-term/sqrt(2*pi*z))
                series.append(Order(1/z**n, x))
                return exp(z)*(Add(*series))

        # All other points may not be handled as of now
        return super(self)._eval_aseries(n, args0, x, logx)
```

But it won't be of much use to us and we won't be able to solve any limit issues mentioned in the pr template through it .
Now the link you've pointed out
<img width="368" alt="image" src="https://user-images.githubusercontent.com/87052487/185842872-3992edb5-7466-47bb-8516-154c4ee3acd6.png">

Is just a breakdown of the `exp(z)` into `sin` and `cos` . Hence we see a form like `sin(z)*Summation1 + cos(z)*Summation2` . At the end it follows the form of aseries for a bessel function which is `exp(z)*(Sum(f(k, nu), (k, 0, oo)))` . 
Hence now I'm afraid that even if we implement this , we would follow the same footsteps as we've seen with `exp(1/x)` and end up with a `PoleError` due to `sin(1/x)` or `cos(1/x)` . I'm quite sure that would be the case eventually because of 
```
>>> sin(1/x)._eval_nseries(x, 6, None)
Traceback (most recent call last):
..............
sympy.core.function.PoleError: Cannot expand sin(1/x) around 0
>>>
>>> cos(1/x)._eval_nseries(x, 5, None)
Traceback (most recent call last):
..............
sympy.core.function.PoleError: Cannot expand cos(1/x) around 0

Reply:
Hence I don't think that we would need a full fledged series for solving any of the limits related to bessel functions at infinities . A good asymptotic approximation would suffice .

Reply:
> I fail to find one for bessely.

I gathered that this was referring to the limits and I was hoping that the expansion in DLMF could be used to compute the limits of `bessely` at `I*oo` etc. The first term of the series should probably suffice. I did not expect that the whole series should be implemented. So the question is, what is the limit of `sin(omega)` when `z -> I*oo`.

Reply:
Ahh sorry for missing out on hint there ! Thanks for pointing this , makes more sense to me how we could make use of `sin` here . I'll try out few things in the direction you're suggesting !

Reply:
Not sure , it will help us out . Like if we apply this diff , based on the first term of the series you mentioned.
```
+++ b/sympy/functions/special/bessel.py
@@ -364,7 +364,7 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         elif e.is_negative:
             # Refer Abramowitz and Stegun 1965, p. 364 for more information on
             # asymptotic approximation of bessely function.
-            return sqrt(2)*sin(z - pi*nu/2 - pi/4)/sqrt(pi*z)
+           return sqrt(2)*(-sin(pi*nu/2 - z + pi/4) + 3*cos(pi*nu/2 - z + pi/4)/(8*z))*sqrt(1/z)/sqrt(pi)
```
Now what happens is in limits.py , first the execution goes to line `coeff, ex = newe.leadterm(z, cdir=cdir)` which obviously gives a `ValueError`
```
        try:
            coeff, ex = newe.leadterm(z, cdir=cdir)
        except (ValueError, NotImplementedError, PoleError):
            e = powsimp(e)
            if e.is_Pow:
                r = self.pow_heuristics(e)
                if r is not None:
                    return r
            try:
                coeff = newe.as_leading_term(z, cdir=cdir)
                if coeff != newe and coeff.has(exp):
                    return gruntz(coeff, z, 0, "-" if re(cdir).is_negative else "+")
            except (ValueError, NotImplementedError, PoleError):
                pass
```

Now `newe = bessely(1, I/x)` and `coeff = sqrt(2)*sqrt(-I*x)*(3*I*x*sin(pi/4 - I/x)/8 - cos(pi/4 - I/x))/sqrt(pi)` , now as we're just checking for `exp` terms in the expression  so we won't really compute `gruntz` of `coeff` here but even if we did.
```
>>> gruntz(sqrt(2)*sqrt(-I*x)*(3*I*x*sin(pi/4 - I/x)/8 - cos(pi/4 - I/x))/sqrt(pi), x, 0)
..................
    raise PoleError("Cannot expand %s around 0" % (self))
sympy.core.function.PoleError: Cannot expand sin(pi/4 + I/_w) around 0
```
Also if we stick to the what we have as of now , we get `coeff = -sqrt(2)*cos(pi/4 - I/x)/(sqrt(pi)*sqrt(I/x))` and applying gruntz we get
```
>>> gruntz(-sqrt(2)*cos(pi/4 - I/x)/(sqrt(pi)*sqrt(I/x)), x, 0)
Traceback (most recent call last):
..............................
    raise PoleError("Cannot expand %s around 0" % (self))
sympy.core.function.PoleError: Cannot expand sin(pi/4 + I/_w) around 0
```
Therefore at the end we get to heuristics which can address then if we can find `bessely(n, I*oo)` where n is integer. Something like `besselj`, is never caught up because it uses `besselj(n, I*oo) == I**(n)*besseli(n, oo)` .

In simplest form gruntz can't help us calculate the following but it correctly calculates results of few close relatives of what we want
```
>>> gruntz(cos(pi/4 - I/x)/sqrt(I/x), x, 0)
Traceback (most recent call last):
..............................
    raise PoleError("Cannot expand %s around 0" % (self))
sympy.core.function.PoleError: Cannot expand sin(pi/4 + I/_w) around 0
>>>
>>> gruntz(cos(pi/3 - I/x)/sqrt(I/x), x, 0)
Traceback (most recent call last):
..............................
    raise PoleError("Cannot expand %s around 0" % (self))
sympy.core.function.PoleError: Cannot expand sin(pi/6 + I/_w) around 0
>>>
>>> gruntz(cos(-I/x)/sqrt(I/x), x, 0)
-oo*sign(I**(3/2))
>>> gruntz(cos(pi-I/x)/sqrt(I/x), x, 0)
oo*sign(I**(3/2))
>>> gruntz(cos(pi/2-I/x)/sqrt(I/x), x, 0)
-oo*I*sign(I**(3/2))
```

Reply:
The exponential approximation I was using earlier (from Wolfram) is something which does give us some breakthrough . I was using `(-I*exp(-I*pi*nu/2 + I*z - I*pi/4) + I*exp(I*pi*nu/2 - I*z + I*pi/4))/sqrt(2*pi*z)` . Using this we get `coeff = sqrt(2)*(-I*exp(-3*I*pi/4 - 1/x) + I*exp(3*I*pi/4 + 1/x))/(2*sqrt(pi)*sqrt(I/x))` . Using gruntz on this works
```
>>> gruntz(sqrt(2)*(-I*exp(-3*I*pi/4 - 1/x) + I*exp(3*I*pi/4 + 1/x))/(2*sqrt(pi)*sqrt(I/x)), x, 0)
-oo*I*sign(I**(3/2))
>>> _.n()
-oo
>>> gruntz(sqrt(2)*(-I*exp(-3*I*pi/4 - 1/x) + I*exp(3*I*pi/4 + 1/x))/(2*sqrt(pi)*sqrt(I/x)), x, 0, '-')
-oo*I*sign(1/sqrt(-I))
>>> _.n()
-oo
```

Reply:
Also as `gruntz` was essentially designed to deal with real exp-log functions , I'm not sure if we're demanding more out of it , expecting results for these expressions . But yeah `besseli`, `besselk` have not so complex exponential  approximations and `besselj` has it's relation with `besseli` as follows
```
        if nu.is_integer:
            if nu.could_extract_minus_sign():
                return S.NegativeOne**(-nu)*besselj(-nu, z)
            newz = z.extract_multiplicatively(I)
            if newz:  # NOTE we don't want to change the function if z==0
                return I**(nu)*besseli(nu, newz)
```
Hence these would never face any issues around `I*oo`  or other complex infinites, the only one which might find some is `bessely` (unless and untill we can find out what value would a particular `bessely(n, I*oo))` return. If we can find these values , heuristics can easily do the job !

Reply:
It seems that the limit of `bessely` at `oo` and `-oo` is computed here: https://github.com/sympy/sympy/blob/88664e6e0b781d0a8b5347896af74b555e92891e/sympy/functions/special/bessel.py#L320-L321
Cannot we add similar lines for `I*oo` and `-I*oo`?

Reply:
Yeah correct but like that's what I've pointed out above, I haven't been able to find any reference regarding how could we possibly evaluate this . I surely know there is a must be a well defined formula because of the repeating pattern of results here .

These are expected results from wolfram 
```
>>> bessely(1, I*oo)
-oo
>>> bessely(2, I*oo)
-I*oo
>>> bessely(3, I*oo)
oo
>>> bessely(4, I*oo)
I*oo
>>> bessely(5, I*oo)
-oo
>>> bessely(6, I*oo)
-I*oo
>>> # answer repeat after every 4 integers
```
To me it looks like like `bessely(n, I*oo) = I**(n + 1)*oo`  ( from the but like I am not finding any good resource to confirm this ). I see `-I*oo` working out similarly .

Reply:
I think that those results would come from the first term of the expansion containing `sin(omega`. If we write `omega = x + I*y` where `x` depends on `nu` and `y -> oo` (or `-oo`), then we get `sin(x)*cos(I*y) + cos(x)*sin(I*y) = sin(x)*cosh(y) + cos(x)*I*sinh(y)`. Now `cosh(y)` and `sinh(y)` agree asymptotically (with `exp(x)/2`) and tend to `oo` when `y -> oo` so it looks like the limit should be `(sin(x) + I*cos(x))*oo = I*(-I*sin(x) + cos(x))*oo = I*exp(-x)*oo` (unless I miscalculated, it should be possible to check numerically).

From the leading factor `sqrt(2/pi*z)` we'll also get and additional factor `sqrt(1/I)` as `z -> I*oo`. (`sqrt(2/pi)` is positive and does not matter. `sqrt(1/y)` will be absorbed by `cosh(y) ~ exp(y)/2 ~ sinh(y)`.)

Reply:
I can confirm the calculations are correct here . So at the end everything boils down to this expression
expr = `I*sqrt(-I)*exp(-I*(-pi*nu/2 - pi/4)) * oo`  for `z` tending to `I*oo`.

We can also breakdown an expression for `-I*oo` similarly. There we would be using the fact the `cosh(-y)` , `sinh(-y)` and `exp(-y)` all agree asymptotically to be `oo` at `y -> -oo`.
Therefore if we add a check like this
```
        if z in (S.Infinity, S.NegativeInfinity):
            return S.Zero
        if z is S.ImaginaryUnit*S.Infinity:
            coeff = I*sqrt(-I)*exp(-I*(-pi*nu/2 - pi/4))
            return coeff.n() * S.Infinity
        if z is -S.ImaginaryUnit*S.Infinity:
            coeff = sqrt(-I)*exp(I*(-pi*nu/2 - pi/4))
            return coeff.n() * S.Infinity  
```

We get all possible correct results
```
>>> bessely(1, I*oo)
-oo
>>> bessely(2, I*oo)
-oo*I
>>> bessely(3, I*oo)
oo
>>> bessely(4, I*oo)
oo*I
>>> bessely(1, -I*oo)
-oo
>>> bessely(2, -I*oo)
oo*I
>>> bessely(3, -I*oo)
oo
>>> bessely(4, -I*oo)
-oo*I
```
I made sure to check all these results against wolfram.